pkg/streamingpromql/engine_fuzz_test.go

```diff
@@ -0,0 +1,234 @@
+}
+
+// This test function will run like a normal unit test with go test
+// It can be used for Fuzz testing with go test -fuzz=FuzzQuery
+// These tests can be extended in 2 ways;
+// a. Add more sample data to the seed-data.test file
+// b. Add more sample queries to the seed-queries.test file
+// The Go Fuzzer will create new queries based off the entries in seed-queries.test
+func FuzzQuery(f *testing.F) {
+
+	// seed some test data - see prometheus/promql/promqltest/README.md
+	data, err := os.ReadFile("testdata/fuzz/data/seed-data.test")
+	require.NoError(f, err)
+
+	opts := NewTestEngineOpts()
+	mimirEngine, err := NewEngine(opts, NewStaticQueryLimitsProvider(0), stats.NewQueryMetrics(nil), NewQueryPlanner(opts), log.NewNopLogger())
+	require.NoError(f, err)
+
+	prometheusEngine := promql.NewEngine(opts.CommonOpts)
+
+	testStorage := promqltest.LoadedStorage(f, string(data))
+	f.Cleanup(func() { require.NoError(f, testStorage.Close()) })
+
+	// Seed some example queries/expressions
+	queries, err := os.ReadFile("testdata/fuzz/data/seed-queries.test")
+	require.NoError(f, err)
+	for _, query := range strings.Split(string(queries), "\n") {
+		f.Add(query)
+	}
+
+	f.Fuzz(func(t *testing.T, query string) {
+		DoComparisonQuery(t, true, query, prometheusEngine, mimirEngine, testStorage)
+		DoComparisonQuery(t, false, query, prometheusEngine, mimirEngine, testStorage)
+	})
+}
+
+// Test different combinations of start, end times in a range query
+// Note that we do not compare to the Prometheus engine, as it validates range inputs in it's http api handler
+func FuzzRangeQueryTimes(f *testing.F) {
+
+	const expression = "vector(0)"
+	const duration = time.Duration(60*5) * time.Second
+
+	opts := NewTestEngineOpts()
+	mimirEngine, err := NewEngine(opts, NewStaticQueryLimitsProvider(0), stats.NewQueryMetrics(nil), NewQueryPlanner(opts), log.NewNopLogger())
+	require.NoError(f, err)
+
+	now := time.Now()
+
+	testStart := []time.Time{now.Add(-time.Hour), now.Add(-time.Second), now.Add(-time.Millisecond), now.Add(-time.Hour * 24 * 365), now, now.Add(time.Second)}
+	for _, start := range testStart {
+		f.Add(start.UnixMilli(), now.UnixMilli())
+	}
+
+	f.Fuzz(func(t *testing.T, startEpoch int64, endEpoch int64) {
+
+		start := time.UnixMilli(startEpoch)
+		end := time.UnixMilli(endEpoch)
+
+		_, mimErr := mimirEngine.NewRangeQuery(context.Background(), nil, nil, expression, start, end, duration)
+
+		if startEpoch <= endEpoch {
+			require.Nil(t, mimErr, "Expected success for start=%s, end=%s", start, end)
+		} else {
+			require.NotNil(t, mimErr, "Expected failure for start=%s, end=%s", start, end)
+		}
+	})
+}
+
+// Seed the fuzz with sample durations in a range query
+// These tests we only check the Mimir engine, as the step size is validated in the Prometheus API handler (api.go)
+// Note also that the max resolution size (end-start/step) is validated in codec.go so we don't test that here
```

Comment:
Might be worth clarifying if the purpose of this is to check the correctness of queries with different steps, or just the correctness of validation of the step and query range. (The answer to this might invalidate some of my other comments for this test.)